### PR TITLE
Automatic item and node colorization

### PR DESCRIPTION
**Pull request by juhdanad**
Now nodes with a palette yield colored item stacks, and colored items place colored nodes by default.

The client also predicts the colorization, so players will immediately see the placed node in the correct color.

### CODE REVIEW

**Comment by juhdanad** [comment]
This pull in itself won't work in creative mode, it requires https://github.com/minetest/minetest_game/pull/1707 too.

File: builtin/game/item.lua [hunk 1]
```diff
@@ -155,12 +155,27 @@ function core.yaw_to_dir(yaw)
 	return {x = -math.sin(yaw), y = 0, z = math.cos(yaw)}
 end
 
-function core.get_node_drops(nodename, toolname)
-	local def = core.registered_nodes[nodename]
+function core.get_node_drops(node, toolname)
+	local def = core.registered_nodes[node.name]
```
Comment (SmallJoker) [inline builtin/game/item.lua]: Needs backwards compability when this argument is a string.

**Comment by juhdanad** [comment]
@SmallJoker Done!

File: builtin/game/item.lua [hunk 2]
```diff
@@ -286,6 +301,25 @@ function core.item_place_node(itemstack, placer, pointed_thing, param2)
...
+	if metatable.palette_index and not def.place_param2 then
+		local color_divisor = 0
+		if def.paramtype2 == "color" then
+			color_divisor = 1
+		elseif def.paramtype2 == "colorwallmounted" then
+			color_divisor = 8
+		elseif def.paramtype2 == "colorfacedir" then
+			color_divisor = 32
+		end
+		if color_divisor > 0 then
```
Comment (VanessaE) [inline builtin/game/item.lua]: set this to nil on line 308 and do `if color_divisor then` here.

**Comment by juhdanad** [comment]
@VanessaE Done!

**Comment by paramat** [comment]
https://github.com/minetest/minetest_game/pull/1707 merged.

**Comment by juhdanad** [comment]
What about this? https://github.com/minetest/minetest_game/pull/1707 won't work without this pull merged.

**Comment by paramat** [comment]
Then this should be in release.

File: builtin/game/item.lua [hunk 3]
```diff
@@ -155,12 +155,35 @@ function core.yaw_to_dir(yaw)
 	return {x = -math.sin(yaw), y = 0, z = math.cos(yaw)}
 end
 
-function core.get_node_drops(nodename, toolname)
+function core.get_node_drops(node, toolname)
+	-- Compatibility, if node is string
+	local nodename = node
+	local param2 = 0
+	-- New format, if node is table
+	if(type(node) == "table") then
```
Comment (paramat) [inline builtin/game/item.lua]: space after 'if'

File: doc/lua_api.txt [hunk 4]
```diff
@@ -3110,8 +3110,10 @@ An `InvRef` is a reference to an inventory.
 * `add_item(listname, stack)`: add item somewhere in list, returns leftover `ItemStack`
 * `room_for_item(listname, stack):` returns `true` if the stack of items
   can be fully added to the list
-* `contains_item(listname, stack)`: returns `true` if the stack of items
-  can be fully taken from the list
+* `contains_item(listname, stack, [match_meta])`: returns `true` if
+  the stack of items can be fully taken from the list.
+  If `match_meta` is false, only the items' names are compared.
+  The default is false.
```
Comment (paramat) [inline doc/lua_api.txt]: Clearer and shorter to have the statement of default in the same sentence as 'match meta'
` If `match_meta` is false, only the items' names are compared (default:false).`

**Comment by paramat** [comment]
Notifying reviewers of previous colour code @sofar @sfan5 @Ekdohibs @Zeno-

**Comment by nerzhul** [comment]
@paramat blocker is for issues, not PR

**Comment by paramat** [comment]
Hmm i'm not sure, what about if there is a PR, with no matching issue, that must be added for release, such as this?

**Comment by juhdanad** [comment]
@paramat Your requests are done! Would you like a small mod for testing?

**Comment by nerzhul** [comment]
@juhdanad adding a single usage of this feature in minimal is a good step to test it (minimal is intended to have many single feature to make them easy to debug)

**Comment by paramat** [comment]
'Blocker' label is useful to flag PRs that must be merged before release, we then do not need to refer to issues to see which PRs need merging.

**Comment by nerzhul** [comment]
No. else i set blocker on all PR i want to see in the release.

**Comment by paramat** [comment]
>  i set blocker on all PR i want to see in the release.

'What a dev wants to see in release' is not what 'blocker' is used for.
So where is the blocker issue for this PR? WIll you add one?

**Comment by nerzhul** [comment]
there is no issue here, it's a new feature.

**Comment by paramat** [comment]
Wrong.

> What about this? minetest/minetest_game#1707 won't work without this pull merged.

The 2 are interdependent, MT is broken unless both are merged, therefore this is a blocking issue.

**Comment by paramat** [comment]
Using 'blocker' in PRs as we have done before makes dev easier because we can then filter PRs to see what has to be merged, instead of filtering issues and then following a link to the PR.
You have decided a change of use for no reason which makes dev harder.

**Comment by nerzhul** [comment]
We have milestones and project for that now on github and it's used.

**Comment by paramat** [comment]
A milestone is not the same as a blocker, a milestone is what should be considered for release but is not absolutely essential, a blocker is what must be included.

**Comment by paramat** [comment]
The project feature does not add labels to the thread list.

File: src/script/lua_api/l_inventory.cpp [hunk 5]
```diff
@@ -325,17 +325,20 @@ int InvRef::l_room_for_item(lua_State *L)
...
 {
 	NO_MAP_LOCK_REQUIRED;
 	InvRef *ref = checkobject(L, 1);
 	const char *listname = luaL_checkstring(L, 2);
 	ItemStack item = read_item(L, 3, getServer(L)->idef());
 	InventoryList *list = getlist(L, ref, listname);
+	bool match_meta = false;
+	if (lua_isboolean(L, 4))
+		match_meta = lua_toboolean(L, 4);
 	if(list){
```
Comment (nerzhul) [inline src/script/lua_api/l_inventory.cpp]: please fix this brace as you modified code around it

**Comment by paramat** [comment]
@juhdanad The MTG PR was merged before this PR, which was the wrong way around, my mistake. Could you explain whether the 2 PRs are essential for release to finish the hardware colouring feature? Because either this is merged or the MTG PR is removed.

**Comment by paramat** [comment]
Unfortunately i cannot give an official +1 as i have almost no understanding of hardware colourisation, it's up to those with some knowledge to review, those i tagged above.

**Comment by paramat** [comment]
I'm not insisting on this but that should be `if (list) {`

**Comment by juhdanad** [comment]
@paramat thanks for noticing! I amended the last commit.

**Comment by juhdanad** [comment]
For testing: this is @sofar's [PoC wool mod](https://gist.github.com/sofar/3c0f9bffff145bced756d7fa771c787a), which does not use this PR (and does not work, because `paletteindex` was renamed to `palette_index`). And [this](https://gist.github.com/juhdanad/433a0aa7094ae0dceaf00924d6fd1e3b) is the modified version for this pull. You can see that the custom `on_place` and `on_dig` callbacks are not needed anymore.
I also removed the custom `description` field, because it would disappear when placing the node (so all nodes are named as "wool", but they have different colors).

**Comment by nerzhul** [comment]
Okay for me, server owners should be careful about retro compat on this mod

**Comment by juhdanad** [comment]
I have added documentation to match with #5876.

**Comment by Zeno-** [comment]
:+1:

**Pull request merged**